### PR TITLE
Rigol MSO5000 support for digital channels in memory mode

### PR DESCRIPTION
**Pull request by jr-oss**
This PR implements reading of digital channels in memory mode from Rigol MSO5000
It addresses bug #1609 (MSO5000) and the implementation might also be fit for bug #1191 (MSO1104Z)
Please see commit messages for details

It still has issues with sample rate and digital-only mode.

### CODE REVIEW

**Comment by brainstorm** [comment]
Hey, [as discussed, thanks a ton for this patch indeed](https://github.com/jr-oss/libsigrok/commit/f806d14bc7562c1f394f7e0030b8a60bf08c0e6a#commitcomment-42603254)!

It seems to work on live capture mode, but I'm not sure it does on memory nor segmented mode (nothing is shown post acquisition)... I think that at least the memory mode would be very valuable to have since by the time it has acquired the 16th LA channel, the first channel is way behind/ahead :-S ... hope that makes sense? Still getting used to how those modes work on my scope.

**Comment by jr-oss** [comment]
Hi, I concentrated too much on logic-only testing, that I did not notice a fatal error in analog channel capturing.
It does not stop at the end of the memory.
I am working on a fix. Currently it looks like this
--- a/src/hardware/rigol-ds/protocol.c
+++ b/src/hardware/rigol-ds/protocol.c
@@ -683,8 +683,13 @@ SR_PRIV int rigol_ds_receive(int fd, int revents, void *cb_data)
        ch = devc->channel_entry->data;
 
        if (ch->type == SR_CHANNEL_ANALOG) {
-               expected_data_bytes = devc->block_start_offset_current_iteration == 0 ?
-                               devc->analog_frame_size : 0;
+               if (devc->block_start_offset_current_iteration == 0) {
+                       expected_data_bytes = devc->analog_frame_size;
+                       offset_stop = MIN(devc->num_channel_bytes + ACQ_BLOCK_SIZE, expected_data_bytes);
+               } else {
+                       expected_data_bytes = 0;
+                       offset_stop = 0;
+               }
        } else {
                expected_data_bytes = devc->digital_frame_size;
 
@@ -697,6 +702,8 @@ SR_PRIV int rigol_ds_receive(int fd, int revents, void *cb_data)
                                devc->frame_needs_another_iteration = TRUE;
                        }
                }
+               offset_stop = devc->num_channel_bytes +
+                               devc->block_start_offset_current_iteration + expected_data_bytes;
        }
 
 
@@ -706,7 +713,7 @@ SR_PRIV int rigol_ds_receive(int fd, int revents, void *cb_data)
        } else if (devc->num_block_bytes == 0) {
                if (devc->model->series->protocol >= PROTOCOL_V4) {
                        offset_start = devc->num_channel_bytes + devc->block_start_offset_current_iteration + 1;
-                       offset_stop = offset_start - 1 + MIN(expected_data_bytes, ACQ_BLOCK_SIZE);
+
                        if (first_frame && rigol_ds_config_set(sdi, ":WAV:START %d",offset_start) != SR_OK)
                                return TRUE;
                        if (first_frame && rigol_ds_config_set(sdi, ":WAV:STOP %d", offset_stop) != SR_OK)

Sorry for not noticing this error before opening the PR

**Comment by jr-oss** [comment]
brainstorm, could you please check, whether your inconsistent data matches bug #1616 ?

**Comment by brainstorm** [comment]
> Hi, I concentrated too much on logic-only testing, that I did not notice a fatal error in analog channel capturing.
> It does not stop at the end of the memory.
> I am working on a fix. Currently it looks like this
> --- a/src/hardware/rigol-ds/protocol.c
> +++ b/src/hardware/rigol-ds/protocol.c
> @@ -683,8 +683,13 @@ SR_PRIV int rigol_ds_receive(int fd, int revents, void *cb_data)
> ch = devc->channel_entry->data;
> (...)
> Sorry for not noticing this error before opening the PR

Sorry for not replying this thread earlier! I'll try to be more diligent and test your changes faster, I would really like to see this PR merged ;)

Your changes look great, exciting to see those bugs ironed... I'll be writing what I'm finding as I go...

I pulled in your changes from the this PR branch (not the `_live` branch yet). Right before acquisition I'm noticing the following messages on the terminal where I launched PulseView:

```
$ ./pulseview
"Note for device developers: Ignoring device configuration capability 'Sample rate' as it is missing GET and/or SET"
"Note for device developers: Ignoring device configuration capability 'Horizontal trigger position' as it is missing GET and/or SET"
"Note for device developers: Ignoring device configuration capability 'Number of horizontal divisions' as it is missing GET and/or SET"
"Note for device developers: Ignoring device configuration capability 'Number of vertical divisions' as it is missing GET and/or SET"
"Note for device developers: Ignoring device configuration capability 'Number of vertical divisions' as it is missing GET and/or SET"
"Note for device developers: Ignoring device configuration capability 'Sample rate' as it is missing GET and/or SET"
"Note for device developers: Ignoring device configuration capability 'Horizontal trigger position' as it is missing GET and/or SET"
"Note for device developers: Ignoring device configuration capability 'Number of horizontal divisions' as it is missing GET and/or SET"
"Note for device developers: Ignoring device configuration capability 'Number of vertical divisions' as it is missing GET and/or SET"
"Note for device developers: Ignoring device configuration capability 'Number of vertical divisions' as it is missing GET and/or SET"
```

Then on acquisition time, with both Analog channels enabled (CH1/CH2) and LA ones (D0/D1) with no signal connected I'm getting the following when stopping the capture:

```
Current logic segment out of range for signal "D0" : 1
Current logic segment out of range for signal "D1" : 1
Current analog segment out of range for signal "CH2" : 1
Current logic segment out of range for signal "D0" : 1
Current logic segment out of range for signal "D1" : 1
Current analog segment out of range for signal "CH2" : 1
Current logic segment out of range for signal "D0" : 1
Current logic segment out of range for signal "D1" : 1
Current analog segment out of range for signal "CH2" : 1
Current logic segment out of range for signal "D0" : 1
Current logic segment out of range for signal "D1" : 1
Current logic segment out of range for signal "D0" : 2
Current logic segment out of range for signal "D1" : 2
Current analog segment out of range for signal "CH2" : 2
Current logic segment out of range for signal "D0" : 2
Current logic segment out of range for signal "D1" : 2
Current analog segment out of range for signal "CH2" : 2
Current logic segment out of range for signal "D0" : 2
Current logic segment out of range for signal "D1" : 2
Current analog segment out of range for signal "CH2" : 2
Current logic segment out of range for signal "D0" : 2
Current logic segment out of range for signal "D1" : 2
Acquisition took 11.14 s
```

Unrelated issue that might be worth fixing... if the oscilloscope gets disconnected from USB and another acquisition is attempted after reboot, PulseView crashes since it cannot find the USB device it was connected to, perhaps a re-enumeration and re-connect would be better instead of crashing:

```
sr: scpi_usbtmc: USBTMC bulk out transfer error: LIBUSB_ERROR_NO_DEVICE.
sr: session: Could not start rigol-ds device usbtmc/20.11 acquisition.
sr: session: Cannot remove non-existing event source 0x7f890b5785f0.
Notifying user of session error: "generic/unspecified error"
2020-10-18 16:37:17.568 pulseview[41096:7822108] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'NSWindow drag regions should only be invalidated on the Main Thread!'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff396e2b57 __exceptionPreprocess + 250
	1   libobjc.A.dylib                     0x00007fff7252c5bf objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff3970b34c -[NSException raise] + 9
	3   AppKit                              0x00007fff369055ec -[NSWindow(NSWindow_Theme) _postWindowNeedsToResetDragMarginsUnlessPostingDisabled] + 310
	4   AppKit                              0x00007fff368ed052 -[NSWindow _initContent:styleMask:backing:defer:contentView:] + 1416
	5   AppKit                              0x00007fff36acddb5 -[NSPanel _initContent:styleMask:backing:defer:contentView:] + 50
	6   AppKit                              0x00007fff368ecac3 -[NSWindow initWithContentRect:styleMask:backing:defer:] + 42
	7   AppKit                              0x00007fff36acdd6a -[NSPanel initWithContentRect:styleMask:backing:defer:] + 64
	8   AppKit                              0x00007fff36c25a28 -[NSWindow initWithContentRect:styleMask:backing:defer:screen:] + 52
	9   libqcocoa.dylib                     0x00000001125813d5 qt_plugin_instance + 181653
	10  libqcocoa.dylib                     0x000000011256dd3d qt_plugin_instance + 102141
	11  libqcocoa.dylib                     0x0000000112567866 qt_plugin_instance + 76326
	12  libqcocoa.dylib                     0x0000000112567340 qt_plugin_instance + 75008
	13  QtGui                               0x000000010ef247e2 _ZN14QWindowPrivate6createEby + 146
	14  QtWidgets                           0x000000010e966703 _ZN14QWidgetPrivate6createEv + 1187
	15  QtWidgets                           0x000000010e965358 _ZN7QWidget6createEybb + 328
	16  QtWidgets                           0x000000010e977d8c _ZN14QWidgetPrivate10setVisibleEb + 428
	17  QtWidgets                           0x000000010eb3e662 _ZN7QDialog10setVisibleEb + 258
	18  QtWidgets                           0x000000010eb3ded3 _ZN7QDialog4execEv + 243
	19  pulseview                           0x000000010d9a99fd _ZN2pv10MainWindow18show_session_errorE7QStringS1_ + 333
	20  pulseview                           0x000000010d9af69a _ZNSt3__110__function6__funcIZN2pv10MainWindow19on_run_stop_clickedEvE3$_1NS_9allocatorIS4_EEFv7QStringEEclEOS7_ + 74
	21  pulseview                           0x000000010d9bd8ae _ZN2pv7Session18sample_thread_procENSt3__18functionIFv7QStringEEE + 1950
	22  pulseview                           0x000000010d9c2298 _ZNSt3__114__thread_proxyINS_5tupleIJNS_10unique_ptrINS_15__thread_structENS_14default_deleteIS3_EEEEMN2pv7SessionEFvNS_8functionIFv7QStringEEEEPS8_SC_EEEEEPvSH_ + 152
	23  libsystem_pthread.dylib             0x00007fff738d9109 _pthread_start + 148
	24  libsystem_pthread.dylib             0x00007fff738d4b8b thread_start + 15
)
libc++abi.dylib: terminating with uncaught exception of type NSException
Abort trap: 6
```

After this, I setup two signals of 5 and 10Hz for both analog and digital channels but live acquisition does not show the digital channels on PulseView (they show up with not problems in the scope):

Skärmavbild 2020-10-18 kl  16 42 03

But then I realise that the trigger should be set to analog CH1 instead of D0 as it is by default on Pulseview:

Skärmavbild 2020-10-18 kl  16 44 48

Changing from `Live` to `Memory` does not seem to work but I'm guessing that you are working on it since you were pasting the diff in here?

I'll tackle [sigrok bugzilla issue 1616](https://sigrok.org/bugzilla/show_bug.cgi?id=1616) next...

**Comment by jr-oss** [comment]
The PR includes the following features for (at least) MSO5000
- get sample rate when using logic channels only
- adjust different sample rates between analog and digital channels
- get correct sample rate, even when switching between Live and Memory data sources
- display logic channels correctly instead of concatenating all channels on D0
- fix memory leaks
- fix re-enabling logic channels, once disabled in PV

Known issues:
- The last commit will probably not work properly when using with sigrok-cli instead of PulseView
[rigol-ds: When the trigger state is already Stop when starting capture, then read that data first before re-arming]
- Reading from memory takes a lot of time. It can be accelerated (17x), when connection via LAN, by setting ACQ_BLOCK_SIZE and ACQ_BUFFER_SIZE to e.g. 1024000

**Comment by brainstorm** [comment]
Thanks Ralf!

Thinking about this one:

> Reading from memory takes a lot of time. It can be accelerated (17x), when connecting via LAN, by setting ACQ_BLOCK_SIZE and ACQ_BUFFER_SIZE to e.g. 1024000

Would it be possible to expose it via env vars, `.sigrokrc` or similar so there are reasonable defaults (not MAX) in addition to a way for users to fine tune their acquisition settings without recompiling?

W.r.t this one:

> Segmented data source does not work. But this seems to be a bug in MSO5000

What are the symptoms? No response/hangs/something else?

**Comment by jr-oss** [comment]
Hi Roman,

Thanks for looking into the PR.
It always helps to get a second opinion and different use cases.

For the non-functioning segmented mode it looks like the scope does not update internal memory properly.
The data is repeated after a number of samples, that could be related to the number of requested bytes, as set with :WAV:START and :WAV:STOP.
I just tried with ACQ_BUFFER_SIZE=1024000 and got repeatedly the first 512000 bytes of the data, no matter what the start value is.

Might still be an issue with my implementation, but on this part I haven't changed that much.


For setting ACQ_BUFFER_SIZE, I would like to discuss with the maintainers about possible ways to implement this.
For USB it does not make sense to increase the value, as the USB scpi interface limits it to 2048 (scpi_usbtmc_lib.c: #define MAX_TRANSFER_LENGTH 2048).
It might be possible to query the scpi interface type and set the buffer size accordingly. 
But at the moment I am only collecting ideas.

**Comment by brainstorm** [comment]
@gsigh @abraxa Can we do something else to review this PR further? Happy to test whatever is needed to get this merged :)

**Comment by vke** [comment]
Are there any chances this PR will be merged? Ready to test and write bug reports

**Comment by abraxa** [comment]
I don't have any of the scopes covered by this driver, so just to clarify - do you expect any regressions for models other than then MSO5000?

**Comment by jr-oss** [comment]
> Am 21.02.2021 um 23:15 schrieb Soeren Apel <notifications@github.com>:
> 
> ﻿
> I don't have any of the scopes covered by this driver, so just to clarify - do you expect any regressions for models other than then MSO5000?
> 
The commits up to 5ed02c27f3c6a852593c12a5029c7cb44fe3cdf4 (including) should cause no regression and provide support for mso5000 with similar limitations as the other scopes and poor (but usable) user experience for logic channels.

Commit afc4722a30f2e6d484d47c4f4b240de142b29753 is hard to judge just by code review.
Here I would prefer if someone could test it.

The remaining commits are mainly cosmetics and could be skipped for first integration.

Last commit b2ac4afd63d24c643d1297bfe49b1e49dc1182b5 will cause regressions, because it is based on my expectations. But I already learned, that others may have different expectations. I still need to provide some discussion page.

If it helps, I can split the PR 

Thanks,
Ralf

**Comment by jr-oss** [comment]
I reverted the commit, that causes a regression (b2ac4af), added a minor bug fix for my modifications and did some code style improvement (c-style comments).

I also made a comparison of debug output (-l 5) between current mainline (libsigrok 0.6.0-git-389acdd9) and current HEAD of this PR (libsigrok 0.6.0-git-210bef14) while capturing 4 analog channels.
The output differs only by timing and modifications from this PR, which are for MSO5000 only or bug fixes.

Thus I assume, that there is no regression in this PR, though I am still in favor of a test with one of the other devices.
I can also clean-up the PR, but did not want to force push without a clear go for it.

**Comment by tdjastrzebski** [comment]
I will be more than happy to test with my MSO5074 when the time comes. Hopes are high :)

**Comment by brainstorm** [comment]
I wish there was somebody reading this PR with a non-MSO5000 scope to test so that the maintainers can merge it with some peace of mind.

Would it be enough to test it with another MSO model from Rigol? Older model, newer?

What are the actual/standard test expectations here, @abraxa?

**Comment by lu94sca** [comment]
Will this ever be merged into master? 
I'm impatiently waiting for my MSO5204 to work with sigrok.

**Comment by jr-oss** [comment]
This PR seems difficult to merge for the maintainers:
- only tested on Rigol MSO5000 scope (few users only AFAIK)
- no guarantees, that I haven't introduced some regression for the already existing/working models. The code already used many if-statements to tweak for the different models and I added some more ;-)
- because it is not in mainline and branches are not built automatically, users cannot contribute easily by testing on other models

If there is anything from my side, that can help to resolve this deadlock, I am happy to contribute and rework my PR

Compiling for Linux is not that difficult (for Windows it is more effort).
You can follow the instructions (http://sigrok.org/gitweb/?p=libsigrok.git;a=blob;f=HACKING;hb=HEAD)

**Comment by lu94sca** [comment]
Unfortunately I'm on Windows.
I tried compiling it following the instructions but I've encountered countless errors and I gave up.
If someone could compile it for me I would appreciate it very much.

**Comment by jr-oss** [comment]
I made another branch for this PR based on current upstream HEAD revision
https://github.com/jr-oss/libsigrok/tree/rigol-ds_mso5000_upstream_head

File: src/hardware/rigol-ds/api.c [hunk 1]
```diff
@@ -1034,8 +1030,8 @@ static int dev_acquisition_start(const struct sr_dev_inst *sdi)
 		break;
 	}
 
-	if (devc->data_source == DATA_SOURCE_LIVE && protocol != PROTOCOL_V5)
-		if (rigol_ds_config_set(sdi, ":RUN") != SR_OK)
+	if (devc->data_source == DATA_SOURCE_LIVE)
+		if (rigol_ds_config_set(sdi, ":SING") != SR_OK)
```
Comment (penguin42) [inline src/hardware/rigol-ds/api.c:1034]: This is making a lot of sense to me; for a start I was seeing retriggering in :RUN so I think multiple channels were showing inconsistent results; but the other effect here is that you select 'live' grab a frame and look at it quickly, and if you like it then you select 'memory' or 'segment' and this means you *don't* retrigger - you just get the data you previously got in live in a high res fetch.
(Although I'm finding hte timescale there is very screwy in segmented mode)

**Comment by nm17** [comment]
Is this PR dead now that jr-oss no longer developing it? Is there anyone developing similar functionality for the MSO5000?

**Comment by abraxa** [comment]
It'll be merged eventually, it's just challenging to review since rigol-ds is one of our most important (and complex) drivers.

**Comment by jr-oss** [comment]
> Is this PR dead now that jr-oss no longer developing it? Is there anyone developing similar functionality for the MSO5000?

No need for changes. MSO5000 just works for me.

But I cannot test with other devices and I agree with abraxa, that the code base is complex and thus this PR is challenging.
I am not very familiar with the code for other model, but tried(tm) not to change anything in that area.
Maybe I find some time to rebase this PR (like [here](https://github.com/jr-oss/libsigrok/tree/rigol-ds_mso5000_upstream_head) ) and test it with my device.
@abraxa Would it be ok to force-push?

**Comment by abraxa** [comment]
You can force-push on branches all you want, yeah.

**Comment by nm17** [comment]
@jr-oss Thank you for your response! I didn't know that the code is complete. I don't remember why, but (I think) I tried your source code a long time ago and I couldn't get it to run on my MSO5000. Also, I thought that you aren't interested in this PR anymore.

Anyways, I can offer any help I need. I'm still very interested in this PR and also #96 . I have an unlocked MSO5000 and a logic probe, so I will test your code tomorrow and get back with you about the results.

> Maybe I find some time to rebase this PR (like [here](https://github.com/jr-oss/libsigrok/tree/rigol-ds_mso5000_upstream_head) ) and test it with my device.

About that: please rebase it if you can. I can also try, but I probably won't be successful because of my lack of familiarity with the libsigrok code base. Do not hesitate to tag me if I can be of any help.

Sorry if I the message I sent before sounded needy or rude, I didn't mean to.
It would really make my year if this PR gets merged into the mainline.
Thank you for your efforts, 4 years later! :heart: :heart: :heart:

**Comment by jr-oss** [comment]
> @jr-oss Thank you for your response! I didn't know that the code is complete. I don't remember why, but (I think) I tried your source code a long time ago and I couldn't get it to run on my MSO5000. Also, I thought that you aren't interested in this PR anymore.

You are welcome.
I am still interested in getting the PR merged, but I also know how difficult it is to judge whether it breaks other devices.
BTW: My MSO5000 runs firmware 00.01.03.00.03. I once tried something newer, that crashed my device as soon as I tried to capture data.

> 
> Anyways, I can offer any help I need. I'm still very interested in this PR and also #96 . I have an unlocked MSO5000 and a logic probe, so I will test your code tomorrow and get back with you about the results.

If you could review/confirm/correct the PR to see, if the implementation for non-PROTOCOL_V5 is unchanged, this would give some confidence.
Some changes affects all PROTOCOL_Vx, but are hopefully compatible or cover digital channels only.
You can follow the commits. The messages should give some hints, why I changed the code.

> 
> > Maybe I find some time to rebase this PR (like [here](https://github.com/jr-oss/libsigrok/tree/rigol-ds_mso5000_upstream_head) ) and test it with my device.
> 
> About that: please rebase it if you can. I can also try, but I probably won't be successful because of my lack of familiarity with the libsigrok code base. Do not hesitate to tag me if I can be of any help.

Done
Basically I took my other branch, that I rebased last year, already resolving conflicts.
PR#96 can take a few days to find the time looking into it. But I don't expect any conflicts.

> 
> Sorry if the message I sent before sounded needy or rude, I didn't mean to. It would really make my year if this PR gets merged into the mainline. Thank you for your efforts, 4 years later! ❤️ ❤️ ❤️

Don't worry.
The PR looked abandoned, because I am not trying to closely follow mainline.

**Comment by nm17** [comment]
Current bugs from my testing. Firmware `00.01.03.02.02`, hardware `01.01.000`. Everything was tested using a USB connection with the factory cable.

  - Live data source recording works really slowly. I'm not sure how fast it should be processed, but it takes about 2-3s for a segment to process.
  - When recording in live data source mode, some channels are inverted. For me, it was D8, D10, D11 and D14. They were always 1, even though I the scope screen says all channels are zero (the probe was connected, but not hooked up to anything)
  - Available channel selection doesn't find CH3 and CH4, even though they are unlocked by options.
  - Segmented or memory data sources don't work for me, Most of the time they just produce no samples or they (rarely) hang the oscilloscope. When it hangs, PulseView says `sr: scpi_usbtmc: USBTMC bulk out transfer error: LIBUSB_ERROR_TIMEOUT.`. Oscilloscope needs to be rebooted in order to work again.
  - Live recording LA output with channels is not time scaled properly. See image (the CH1 trace is 4995us and LA trace is 80ms, measured by the cursors tool):
image

**Comment by jr-oss** [comment]
> Current bugs from my testing. Firmware `00.01.03.02.02`, hardware `01.01.000`. Everything was tested using a USB connection with the factory cable.

USB is very sensitive on my device. As soon as I abort a transfer I have to reboot the scope.
Thus I only use the Ethernet connection.

> 
>     * Live data source recording works really slowly. I'm not sure how fast it should be processed, but it takes about 2-3s for a segment to process.

My scope takes about 1.2s for each frame in Live data source. 
You could set loglevel to 5 and check the timing to find the bottleneck
For best performance (in memory data source) I use locally these settings, which are not part of this PR, because it will break other devices: https://github.com/jr-oss/libsigrok/commit/fbd8d51a1eae12db87a2546a6a13111deddffb04


> 
>     * When recording in live data source mode, some channels are inverted. For me, it was D8, D10, D11 and D14. They were always 1, even though the scope screen says all channels are zero (the probe was connected, but not hooked up to anything)

No idea. Can you check the data, that is transferred between scope and PC?
With Ethernet you can use Wireshark to look into the data.
Maybe Rigol changed the firmware to finally match the data sheet ...
See commit message https://github.com/sigrokproject/libsigrok/commit/64c3ae8ea804a22ac89ae2750a1fec5e6177338d

> 
>     * Available channel selection doesn't find CH3 and CH4, even though they are unlocked by options.

If the scope does not have CH3 and CH4 natively, you probably have to tweak this table
https://github.com/jr-oss/libsigrok/blob/master/src/hardware/rigol-ds/api.c#L199

> 
>     * Segmented or memory data sources don't work for me, Most of the time they just produce no samples or they (rarely) hang the oscilloscope. When it hangs, PulseView says `sr: scpi_usbtmc: USBTMC bulk out transfer error: LIBUSB_ERROR_TIMEOUT.`. Oscilloscope needs to be rebooted in order to work again.

Segmented data source does not work for me either, but memory works.
See also: https://sigrok.org/wiki/Rigol-ds_data_source

> 
>     * Live recording LA output with channels is not time scaled properly. See image (the CH1 trace is 4995us and LA trace is 80ms, measured by the cursors tool):

Hmm, this is a factor of 16, like the number of digital channels.
I had to stretch LA signals to adjust sample rates between digital an analog channels.
See https://github.com/sigrokproject/libsigrok/commit/64c3ae8ea804a22ac89ae2750a1fec5e6177338d#diff-77c0ee0dc3269cbc66820e7079da7fc2947153e6cf076c95a609f97ef237cf81R835

**Comment by nm17** [comment]
> USB is very sensitive on my device. As soon as I abort a transfer I have to reboot the scope.

Weird, it's fine for me in that regard.

> Thus I only use the Ethernet connection.

I also want to use it over Ethernet, but I sadly don't have a switch or a second port.

> My scope takes about 1.2s for each frame in Live data source.
> You could set loglevel to 5 and check the timing to find the bottleneck

Will do, thanks for the heads up.

> With Ethernet you can use Wireshark to look into the data.

I think I can record USB data with it, will try to record it and send it to you.

> If the scope does not have CH3 and CH4 natively, you probably have to tweak this table
> https://github.com/jr-oss/libsigrok/blob/master/src/hardware/rigol-ds/api.c#L199

That's certainly not user friendly, this should be done through a option, in my opinion. That should be the easiest thing to do. But I guess that's out of scope of this PR.
